### DISCUSSION TITLE
Memory allocation error when exporting sliced xarray data array to a numpy array

### DISCUSSION THREAD

**Discussion by lassiterdc** [pydata/xarray#5771] [Q&A]
The sliced data array is 0.002 megabytes but when I try to export it to a numpy array, xarray tries to load the entire unsliced data array into memory which causes a memory allocation error (which is in itself odd since the dataset is only 318MB).

MRMS precipitation data in grib2 format (318MB): https://www.dropbox.com/s/xq0vvamcydu3qn2/PrecipRate_20210524.grib2?dl=0

```
# you also need cfgrib installed
# conda install -c conda-forge cfgrib
# https://github.com/ecmwf/cfgrib

#%% Data processing
import xarray as xr

f_in_grib = "PrecipRate_20210524.grib2"

mrms = xr.open_dataset(f_in_grib, engine="cfgrib", chunks = {'latitude':10, 'longitude':10})

idx = dict(latitude=1808, longitude=5374)

data_at_gages = mrms[idx]

#%% inspection
print("The dataset 'data_at_gages' is {} megabytes.".format(data_at_gages.unknown.nbytes/1e6))
# The dataset 'data_at_gages' is 0.0021 megabytes.

print("The dimensions of the sliced data array are: {}".format(data_at_gages.unknown.shape))
# The dimensions of the sliced data array are: (525,)

#%% error
data_at_gages.unknown.values.to_numpy

# MemoryError: Unable to allocate 47.9 GiB for an array with shape (525, 3500, 7000) and data type float32
```

**Comment by mliukis** [comment]
It could be the Dask issue - please take a look at  [my other reported issue](https://github.com/pydata/xarray/discussions/5652). Downgrading Dask took care of the problem for me.

**Reply by lassiterdc** [reply]
Thanks for the suggestion! I just tried that though and am still getting the same error. As a sidenote, instead of chunking based on lat and long I chunked based on time `{chunks  = "time":10}` and my error has changed. Now it says `Unable to allocate 187. MiB for an array with shape (24500000,) and data type float64` which is super weird because my computer has 32GB of RAM.

**Reply by mliukis** [reply]
Hmm... It is probably running out of memory by the time it needs to allocate 187Mb of memory. You can watch process's memory usage with `top` (OS X, Linux) or any other memory monitoring tools available on your platform to see if it indeed consumes all the memory.

**Reply by lassiterdc** [reply]
Ah you are right, it just keeps incrementing up till I cap my memory... Is there a way to avoid that or do I just need tons of RAM to access a time series for a single lat/long?

**Reply by mliukis** [reply]
That what was happening to my process converting huge Zarr store to NetCDF - the memory usage kept climbing though I was using chunked Dask arrays in `xarray`. 
Sorry for not being of much help. From my experience with xarray, it could be the xarray and dependent Python packages interaction that is causing the problem. My other `xarray` debugging experience was that I had to downgrade Zarr package as it was causing corrupted datetime objects in Zarr. Sigh...
I hope some xarray gurus will come to your rescue :) Good luck!

**Reply by mliukis** [reply]
Something I just noticed in the following line:

`#%% error
data_at_gages.unknown.values.to_numpy`

Shouldn't `data_at_gages.unknown.values` be already of the `np.ndarray `type according to [values](https://xarray.pydata.org/en/v0.19.0/generated/xarray.DataArray.values.html)? Then you don't need to call to_numpy, and I would expect you getting a syntax error instead of running out of memory :)
Most likely it's running out of memory accessing `values`

**Reply by lassiterdc** [reply]
Ah okay so it sounds like that memory usage climbing is evidence of some kind of malfunction so I'll keep at it.

Your second comment gave me a glimmer of hope but I tried running `data_at_gages.to_netcdf()` and ran into the same issue. If it means anything, I was able to run `data_at_gages.to_dask_dataframe()` but any operation causes the same memory usage climb it seems.

**Reply by mliukis** [reply]
Makes me think it's a Dask version issue. Did you try to downgrade Dask to the same version that worked for me: 

```
dask                      2.30.0                     py_0    conda-forge
```
?
